File: ExamRegisterBot.js
const request = require('request');
const cheerio = require('cheerio');
const querystring = require('querystring');

const timer = 5000;
//Insert your cookie here!
const cookie = '';


if (cookie === '')
{
	console.error("Please Insert your cookie!");
	process.exit();
}

function check_exam()
{
	request.get({
	url :"https://profile.intra.42.fr", 
	headers: {
		'Cookie' : cookie, 
	}},
	(err, res, body) => {
		if (err)
		{
			console.error("[-] Check your connection!");
			return ;
		}
		console.log("[+] Connected to the site!");

		let $ = cheerio.load(body);
		let exam_event = $('div[data-event-kind="exam"]').html();
		if (exam_event == null)
		{
			console.log("No EXAM!");
			return ;
		}
		let registered = exam_event.search('<span class="event-registered">registered</span>');
		if (registered !== -1)
		{
			console.log("You are already registered to Exam!");
			process.exit();
		}
		let meta_csrf = $('meta[name="csrf-token"]').attr('content');
		let event_id = exam_event.match(/exams\/[0-9]*/);
		var postData = querystring.stringify({
			_method: 'post',
			authenticity_token: meta_csrf
		})
		request.post({
				url: "https://profile.intra.42.fr/" + event_id[0] + "/exams_users", 
				form : postData,
	 			headers:
				{
					'Accept' : 'text/html,application/xhtml+xml,application/xml;q=0,9,image/webp,image/apng,*/*;q=0.8',
					'Accept-Language': 'en-US,en;q=0.9,fr;q=0.8',
					'Cache-Control': 'max-age=0',
					'Connection': 'keep-alive',
					'Content-Length': postData.length,
					'Content-Type': 'application/x-www-from-urlencoded',
					'Cookie' : cookie,
					'Host'   : 'profile.intra.42.fr',
					'Origin' : 'https://profile.intra.42.fr',
					'Referer': 'https://profile.intra.42.fr/',
					'Upgrade-Insecure-Requests' : '1',
					'User-Agent' : 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.62 Safari/537.36',
				}}, (err, res, body) => {
					if (err)
					{
						console.error("[-] Error:", err);
						return ;
					}
					console.log("[+] Exams user was successfully created.");
					process.exit();
				})
	})
}

setInterval(check_exam, timer);

File: ExamRegisterBot.py
import requests
import re
from time import sleep
from bs4 import BeautifulSoup as bs
from sys import exit

timer = 60
#Insert your cookie!
cookie = ""

def skip_specific_exam(exams):
	for exam in exams:
		if 'class="date-day">15' and 'class="date-month">September' in exam:
			exams.remove(exam)
			# print("Skipping September 15 exam")

def RegisterToExam():
	cookie_dict = {'_intra_42_session_production': cookie}
	try:
		req = requests.get("https://profile.intra.42.fr", cookies=cookie_dict)
		req.raise_for_status()
	except Exception as e:
		print(f"[-] check your connection {e}")
		return
	soup = bs(req.content.decode('utf-8'), 'html.parser')
	exams = [str(x) for x in soup.find_all('div') if x.get('data-event-kind') == 'exam']

	#hmmm there should be an easy way to get the csrf_token
	#maybe just a regex to get the csrf-token is better....

	csrf_token = soup.find(attrs={'name':'csrf-token'})
	if not csrf_token:
		print("[-] cookie is probably invalid")
		exit(1)
	csrf_token = csrf_token['content']

	# skip_specific_exam(exams)

	if not exams:
		# print("[*] no exam")
		return
	for exam in exams:
		# print(f"[*] checking exam: {exam}")
		# continue
		if '<span class>="event full"' in exam:
			continue
		else:
			if '<span class="event-registered">registered</span>' in exam:
				print("[+] already registered to an exam")
				exit(0)
			event_id = re.search('/exams/[0-9]*', exam)
			if event_id:
				post_url = "https://profile.intra.42.fr" + event_id[0] + "/exams_users"
				post_data = {'_method': 'post', 'authenticity_token': csrf_token}
				try:
					req_exam = requests.post(post_url, data=post_data, cookies=cookie_dict)
				except Exception as e:
					print("[-] error:", str(e))
					exit(1)
				print("[+] successfully registered to an exam")
				exit(0)
			else:
				print("[*] contact the author of this program or modify and send a pull request")
	print("every exams are full")

if __name__ == "__main__":
	try:
		while True:
			RegisterToExam()
			sleep(timer)
	except KeyboardInterrupt:
		print("[-] exiting...")
		exit(0)
